### ISSUE TITLE
GP classification with Linear kernel fits poorly depending on initialisation of hyperparameters

### ISSUE THREAD

**Issue by avalonhse** [GPflow/GPflow#1435]
# Bug / performance issue / build issue

I have followed the GPFlow notebook using the VGP model [https://gpflow.readthedocs.io/en/develop/notebooks/basics/classification.html](https://gpflow.readthedocs.io/en/develop/notebooks/basics/classification.html) for the GP Classification method and apply different kernels in a 2-dimensional data. The VGP model works well with Squared Exponential kernel, however, it fails with Linear kernel.

## To reproduce

To reproduce the bug, I have created a Notebook at:
[https://github.com/avalonhse/BayesNotebook/blob/master/Issue_GPFlow_Linear_Classification.ipynb](https://github.com/avalonhse/BayesNotebook/blob/master/Issue_GPFlow_Linear_Classification.ipynb)

In this notebook, I compare 2 models using library GPflow and library GPy for GP classification in a 2-dimensional dataset:
- With the Squared Exponential kernel, both libraries produce almost the same result
- With the Linear kernel, GPflow failed and GPy passed.
- With the Squared Exponential + Linear kernel, GPflow failed and GPy passed.
 
**Minimal, reproducible example**

```python
# This is the place for your code that reproduces the bug.
# Please make sure it does not depend on external libraries (beyond GPflow's own requirements), specific datasets, etc., and the smaller, the better :)
# For help on how to write a good bug report, see https://stackoverflow.com/help/minimal-reproducible-example

import numpy as np
X = np.array([[ 9.96578428, 60.],[ 9.96578428, 40.],[ 9.96578428, 20.],
       [10.96578428, 30.],[11.96578428, 40.],[12.96578428, 50.],
       [12.96578428, 70.],[8.96578428, 30. ],[ 7.96578428, 40.],
       [ 6.96578428, 50.],[ 6.96578428, 30.],[ 6.96578428, 10.],
       [11.4655664 , 71.],[ 8.56605404, 63.],[12.41574177, 69.],
       [10.61562964, 48.],[ 7.61470984, 51.],[ 9.31514956, 45.]])
Y = np.array([[1., 1., 0., 0., 0., 0., 0., 0., 0., 1., 1., 0., 1., 1., 0., 0., 1., 0.]]).T

# plotting
import matplotlib.pyplot as plt
import matplotlib
%matplotlib inline
def plot(X,Y):
    mask = Y[:, 0] == 1

    plt.figure(figsize=(6, 6))
    plt.plot(X[mask, 0], X[mask, 1], "oC0", mew=0, alpha=0.5)

    plt.ylim(-10, 100)
    plt.xlim(5, 15)

    _ = plt.plot(X[np.logical_not(mask), 0], X[np.logical_not(mask), 1], "oC1", mew=0, alpha=0.5)

plot(X,Y)

# Evaluate real function and the predicted probability
res = 500
xx, yy = np.meshgrid(np.linspace(5, 15, res),
                     np.linspace(- 10, 120, res))
Xplot =  np.vstack((xx.flatten(), yy.flatten())).T

# Code followed the Notebook : https://gpflow.readthedocs.io/en/develop/notebooks/basics/classification.html

import tensorflow as tf
import tensorflow_probability as tfp
import gpflow
from gpflow.utilities import print_summary, set_trainable, to_default_float
gpflow.config.set_default_summary_fmt("notebook")

def testGPFlow(k):
    m = gpflow.models.VGP(
        (X, Y), 
        kernel= k,
        likelihood=gpflow.likelihoods.Bernoulli()
    )
    print("\n ########### Model before optimzation ########### \n")
    print_summary(m)
    
    print("\n ########### Model after optimzation ########### \n")
    opt = gpflow.optimizers.Scipy()
    res = opt.minimize(
        m.training_loss, variables=m.trainable_variables, options=dict(maxiter=2500), method="L-BFGS-B"
    )        

    print(' Message: ' + str(res.message) + '\n Status = ' + str(res.status) + '\n Number of iterations = ' + str(res.nit))
    print_summary(m)
    
    means, _ = m.predict_y(Xplot)  # here we only care about the mean
    y_prob = means.numpy().reshape(*xx.shape)

    print("Fitting model using GPFlow")
    plot(X,Y)

    _ = plt.contour(
        xx,
        yy,
        y_prob,
        [0.5],  # plot the p=0.5 contour line only
        colors="k",
        linewidths=1.8,
        zorder=100,
    )

k = gpflow.kernels.Linear(variance= 1)
testGPFlow(k)

```

**Stack trace, or error message**
```
// Paste the full stack trace/error message here
```

## Expected behavior

```python
import GPy 

def testGPy(k):
    m = GPy.models.GPClassification(X,Y,k)
    print("\n ########### Model before optimzation ########### \n",m)
    for i in range(5):
        m.optimize('bfgs', max_iters=50) #first runs EP and then optimizes the kernel parameters
    print("\n ########### Model after optimzation ########### \n",m)
    
    p = m.predict(Xplot)[0].reshape((res, res)) 
    
    print("Fitting model using GPy")
    plot(X,Y)

    _ = plt.contour(
        xx,
        yy,
        p,
        [0.5],  # plot the p=0.5 contour line only
        colors="k",
        linewidths=1.8,
        zorder=100,
    )

k = GPy.kern.RBF(input_dim = 2, variance=7., lengthscale=0.2)
testGPy(k)
```
## System information:

* GPflow version: 2.0.0 
* GPflow installed from: pip install gpflow 
* TensorFlow version:  2.1.0
* Python version 3.6.10 
* Operating system: Ubuntu 18.04

**Comment by st--** [comment]
Hi @avalonhse, thanks for your interest in GPflow and getting in touch with a thorough bug report! I've had a quick think about this, but will have to come back to this later. For now: Classification (or more generally, non-Gaussian likelihoods) using GPs cannot be evaluated exactly, and we need to approximate the posterior. GPy and GPflow use different approximations (Expectation Propagation in GPy's GPClassification, Variational Inference in GPflow's VGP model), so it's not surprising if they don't find *exactly* the same answer, but of course in your example the VGP model seems to not fit at all! I suspect this is due to ending up in a poor local optimum, due to the initialisation and/or the optimisation. For the linear example, note that the GPy model ended up with a fairly low variance (~0.07) whereas in your GPflow example the variance ended up being much higher (~1.7). What happens if you initialise the linear kernel in GPflow with a lower variance (try `gpflow.kernels.Linear(variance=0.1)`)? To improve optimisation, it may be helpful to use natural gradients (see our [GPflow notebook](https://gpflow.readthedocs.io/en/develop/notebooks/advanced/natural_gradients.html)).

**Comment by st--** [comment]
Hi @avalonhse, I've just tried out your notebook, changing the initial variance for the `Linear` kernel from 1.0 to 0.1 does indeed solve your problem - and the VGP fit in the very last example (RBF+Linear) looks even better to me than the GPy EP fit. So I don't think this is any bug inside GPflow, it's just about what you use it for and how you use it. Local optima in maximum likelihood inference are a well-known phenomenon. As such I'm closing this issue now - if you've got further questions on this, it might be more helpful if you start a question on stackoverflow.com (we've got a [gpflow tag](https://stackoverflow.com/questions/tagged/gpflow)), though perhaps the [CrossValidated StatsExchange](https://stats.stackexchange.com/) on statistics might be more appropriate still. If you're unsure how to proceed, please feel free to join the discussion on gpflow.slack.com (invite link is in the GPflow README.md), and you can always re-open issues if need be :)

**Comment by avalonhse** [comment]
Great. Thank you very much.

**Comment by avalonhse** [comment]
Hi @st-- 

I post a follow-up question in stackoverflow.com: https://stackoverflow.com/questions/61333129/gpflow-gp-classification-with-1-dim-linear-kernel-fits-poorly-for-2-dimension .  This question I had before this issue. Much apprciated.
